### PR TITLE
Move `eos_token_id` to stopping criteria

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

This PR is a small step for `torch.compile` and `generate` compatibility. It moves `EOS` token to stopping criteria, so now we can loop `while stopping_criteria` and get rid of extra checks on `EOS` at the end of each generate method. 

All the generate tests, including slow are passing.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

@gante

### CODE REVIEW

File: src/transformers/generation/stopping_criteria.py
Comment (gante) [inline src/transformers/generation/stopping_criteria.py:150]: nit: `torch.isin` is imo cleaner to read ([docs](https://pytorch.org/docs/stable/generated/torch.isin.html))

but highly debatable, feel free to pick the solution you like the most

File: src/transformers/generation/stopping_criteria.py
Comment (gante) [inline src/transformers/generation/stopping_criteria.py:149]: The conversion to tensor can be in `__init__` -- that way we only convert it once, as opposed to once per iteration

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:1941]:
```suggestion
        if eos_token_id is not None:
            warnings.warn(
                "`eos_token_id` is deprecated in this function and will be removed in v4.41, use"
                " `stopping_criteria=StoppingCriteriaList([EOSTokenCriteria(eos_token_id=eos_token_id)])` instead.",
                FutureWarning,
            )
            stopping_criteria.append(EOSTokenCriteria(eos_token_id=eos_token_id))
```
(see 1. my PR-wide comment)
2. If 1. is applied, we don't need to set `eos_token_id`. As such, we don't need the `else` case at all (`eos_token_id` is `None` in the `else`, so the associated stopping criteria shouldn't exist)

File: tests/generation/test_stopping_criteria.py
Comment (gante) [inline tests/generation/test_stopping_criteria.py:111]:
```suggestion
        input_ids, scores = self._get_tensors(5)
        input_ids[:2, -1] = 0
        input_ids[2, -1] = 1
        self.assertListEqual(criteria(input_ids, scores).tolist(), [True, True, False])
```
`_get_tensors` is random, I think we may get failures here if we don't hardcode the test case.

File: tests/generation/test_utils.py
Comment (gante) [inline tests/generation/test_utils.py:3351]: Instead of ignoring the warning, let's make sure `generate` calls doesn't cause warnings :)

**Review by gante — COMMENTED** [review]
Looking good 💪 A few comments to further refine the idea

In the body of `generate`, we set `pad_token_id` to `eos_token_id` when the latter exists and the former is `None`.

As such, we can further modify the decoding functions as follows:
1. the block 
  ```py
              # finished sentences should have their next token be a padding token
              if eos_token_id is not None:
                  if pad_token_id is None:
                      raise ValueError("If `eos_token_id` is defined, make sure that `pad_token_id` is defined.")
                  next_tokens = next_tokens * unfinished_sequences + pad_token_id * (1 - 
  unfinished_sequences)
  ``` 
  can become
  ```py
              # finished sentences should have their next token be a padding token
              if pad_token_id is None:
                  next_tokens = next_tokens * unfinished_sequences + pad_token_id * (1 - unfinished_sequences)
  ``` 
  because the case that raises the exception can never be triggered from `generate` (although we should add an integration test that ensures this remains true, if it doesn't already exist!).

2. see my in-code comment :)

**Comment by gante** [comment]
Ah, the pipeline tests are ignored by default (just like slow tests). You need to add `RUN_PIPELINE_TESTS=1` to run them :)

File: src/transformers/generation/utils.py
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:1941]: Yes, we do not need if the methods are called from `generate`. But I thought of two cases that might lead to errors, because users are still able to call generate methods directly:

1. If they do not pass in "eos" explicitly, neither they pass the new `EOSTokenCriteria`, their generation might not stop at "eos" going until max length. Worse, if they do not pass any StoppingCriteria, then generation will go forever. Before this PR, we always took generation config's "eos" by default and used it to stop generation.
2.  Beam search methods need an "eos" for their scorer's correct functioning. We discussed in one of prev PRs and decided that calling the criterias twice in one loop is not good (but theoretically still possible), so we have to define `eos_token_id` in the beginning anyway. At least until beam search methods are refactored

So the question is, shouldn't we be more cautious until the methods become completely private? Or maybe leave it only for beam search methods, removing completely for other methods?

File: tests/generation/test_stopping_criteria.py
Comment (zucchini-nlp) [reply tests/generation/test_stopping_criteria.py:111]: No, because we hardcoded the last token id and the criteria looks only at the last one. Correct me if I misunderstood you

File: tests/generation/test_utils.py
Comment (zucchini-nlp) [reply tests/generation/test_utils.py:3351]: okay, makes sense, I will remove "eos" from generate calls to avoid deprecation warning 😄

File: src/transformers/generation/stopping_criteria.py
Comment (zucchini-nlp) [reply src/transformers/generation/stopping_criteria.py:149]: no problem, will do so

File: tests/generation/test_stopping_criteria.py
Comment (gante) [reply tests/generation/test_stopping_criteria.py:111]: > the criteria looks only at the last one

Agreed

> we hardcoded the last token id

In this case we hardcode the first two rows (`input_ids[:2, -1] = 0`). The third row has no hardcoded value -- while the corresponding criteria result should be `False` in most cases, nothing (AFAIK) prevents the last token in the last row to be `0`, causing this part of the test to fail.

Comment (zucchini-nlp) [reply tests/generation/test_stopping_criteria.py:111]: got it, accepted

File: src/transformers/generation/utils.py
Comment (gante) [reply src/transformers/generation/utils.py:1941]: You're right, it needs more thought than my code suggestion. Another thing we haven't considered is the case where the user passes both `eos_token_id` and the new stopping criteria.

Let's keep the `else` case with a TODO to remove it after beam search is refactored 👍 Let's also add a case in this if/else: if the new stopping criteria is already present, let's not add a duplicate.

**Comment by zucchini-nlp** [comment]
Done for all comments. I checked that users have to possibility of messing up when calling the these methods directly and ran again all tests (+slow, +pipeline)

File: src/transformers/generation/stopping_criteria.py
Comment (gante) [inline src/transformers/generation/stopping_criteria.py:149]:
```suggestion
        is_done = torch.isin(input_ids[:, -1], self.eos_token_id.to(input_ids.device))
```
(fewer elements to test :))

File: tests/generation/test_utils.py
Comment (gante) [inline tests/generation/test_utils.py:3351]:
```suggestion
```

File: tests/generation/test_utils.py
Comment (gante) [inline tests/generation/test_utils.py:3357]:
```suggestion
```

**Review by gante — APPROVED** [review]
LGTM, thank you for iterating 🙌

File: src/transformers/generation/__init__.py
Comment (ArthurZucker) [inline src/transformers/generation/__init__.py:85]:
```suggestion
        "EosTokenCriteria",
```

Comment (ArthurZucker) [reply src/transformers/generation/__init__.py:85]: I think we should still strive for Camel

File: src/transformers/generation/stopping_criteria.py
Comment (ArthurZucker) [inline src/transformers/generation/stopping_criteria.py:132]:
```suggestion
class EosTokenCriteria(StoppingCriteria):
```

File: src/transformers/generation/stopping_criteria.py
Comment (ArthurZucker) [inline src/transformers/generation/stopping_criteria.py:134]:
```suggestion
    This class can be used to stop generation whenever the "end-of-sequence" token is generated.
```

File: src/transformers/generation/stopping_criteria.py
Comment (ArthurZucker) [inline src/transformers/generation/stopping_criteria.py:135]:
```suggestion
    By default, it uses the `model.generation_config.eos_token_id`.
```

File: src/transformers/generation/stopping_criteria.py
Comment (ArthurZucker) [inline src/transformers/generation/stopping_criteria.py:139]: would be nice to also support list of lists ? To have stopping tokens (if the eos is `["<","eos",">"]`)

File: src/transformers/generation/utils.py
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1923]: It should rather be set in the generation_config I guess? Why can't we leave this as a kwarg and set the generation_config.eos_token_id ?

File: src/transformers/generation/utils.py
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:4671]: this means we the last stopping_criteria is always `EOSTokenCriteria`. It's not necessarily obvious and if people use custom passed criteria I am not sure this will always be the case?

**Review by ArthurZucker — COMMENTED** [review]
Thanks, would be nice to make sure our assumption is always correct, nice cleanup otherwise!

File: tests/generation/test_utils.py
Comment (zucchini-nlp) [reply tests/generation/test_utils.py:3351]: totally forgot to remove ahah

File: src/transformers/generation/stopping_criteria.py
Comment (zucchini-nlp) [reply src/transformers/generation/stopping_criteria.py:139]: Hmm, is it really true for existing models to have eos tokens as a sequence? I guess you are referring to custom eos tokens, when users want to stop at "<eos>" but they haven't trained the model with "<eos>" as special token. If that's the case, there is `StopStringsCriteria` PR coming or users are free to write their custom criteria

@gante wdyt?

File: src/transformers/generation/utils.py
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:4671]: right, will fix it

File: src/transformers/generation/utils.py
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:1923]: That's the general idea, we want to stop accepting `eos` as input argument. The warning is for backward compatibility, since we used to give priority to user defined `eos_token_id` from args before checking the `generation_config.eos_token_id`

File: src/transformers/generation/stopping_criteria.py
Comment (gante) [reply src/transformers/generation/stopping_criteria.py:139]: Yeah, the case where we want to stop on a string will be covered by #28932 (and is way more complex)

This PR is exclusively to port the existing EOS logic into its own stopping criteria :)

File: src/transformers/generation/utils.py
Comment (gante) [reply src/transformers/generation/utils.py:4671]: @ArthurZucker good catch, I missed this one too :)

Comment (zucchini-nlp) [reply src/transformers/generation/utils.py]: @gante need you to look. I removed some code, given that assisted decoding works only with one batch, so we can make some assumptions about when to stop generating.

File: src/transformers/generation/stopping_criteria.py
Comment (ArthurZucker) [reply src/transformers/generation/stopping_criteria.py:139]: Right I forgot it was being covered1

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4637]: 🧠  (this is actually much more versatile than the previous version!)

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:4848]:
```suggestion
        gamma = candidate_logits.shape[1]
```
`n_matches` is the number of correct guesses, `gamma` is the number of total guesses. `min(candidate_logits.shape[1], n_matches)` would always result in `gamma = n_matches`. In turn, we would always be skipping the `if n_matches < gamma` branch below, which is the key ingredient of speculative decoding :) (see [algorithm 1 on page 3](https://arxiv.org/pdf/2211.17192.pdf))

The original `gamma = min(candidate_logits.shape[1], max_matches)` existed to prevent us from going beyond the EOS token, which the updated `is_done_candidate` line should catch

**Review by gante — COMMENTED** [review]
One change in the speculative decoding diff and we're ready to go :)

File: src/transformers/generation/utils.py
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:4848]: right

**Comment by gante** [comment]
@zucchini-nlp needs a rebase with `main` and it should be ready to be merged 

(@ArthurZucker I'm assuming this is ready to be merged, since all comments were addressed. I'm deciding to merge since this unblocks me on the torch.compile front, let us know if you like some post-merge changes :) )

**Comment by amyeroberts** [comment]
@gante Woah, hold on before merging! You still need a core maintainers approval: even if the comments have been addressed it's important to make sure they're amendments are approved

**Comment by gante** [comment]
@zucchini-nlp this PR needs to be rebased with `main` to fix CI :)

**Comment by gante** [comment]
@ArthurZucker ping for approval, if all tasks are complete :)

File: src/transformers/generation/utils.py
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1915]:
```suggestion
            logger.warning_once(
```

File: src/transformers/generation/utils.py
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1917]:
```suggestion
                " `stopping_criteria=StoppingCriteriaList([EosTokenCriteria(eos_token_id=eos_token_id)])` instead.",
                " Otherwise make sure to set `model.generation_config.eos_token_id`"
```

File: src/transformers/generation/utils.py
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:2374]: same comment here

**Review by ArthurZucker — APPROVED** [review]
LGTM!  Make sure to run the slow tests for important models before merging! + testing generation with Llama and eos_token_id list!

**Comment by zucchini-nlp** [comment]
Rebased `main` and ran all tests in generation again, including slow, ran a few generation models' tests. Everything is passing, can be merged now. @gante 

No idea why the examples test is failing, does not have anything to do with this PR.

**Comment by agnosticlines** [comment]
Just a heads up this commit breaks transformers generation on Apple Silicon as isin is not implemented for the MPS backend

**Pull request merged**
